erpnext/erpnext_integrations/doctype/leaflink_settings/leaflink_settings.py

```diff
@@ -0,0 +1,191 @@
+	if not leaflink_settings.get("enable_leaflink"):
+		integration_request.update({
+			"status": "Failed",
+			"error": "Leaflink is not enabled."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+	#validate incoming keys
+	key = bytes(leaflink_settings.get("incoming_api_authentication_key"), 'utf-8')
+	incoming_signature = frappe.local.request.headers['LL-Signature'].encode("utf-8")
+	request_body = frappe.local.request.data
+	computed_signature = base64.b64encode(hmac.new(key, request_body, digestmod=hashlib.sha256).digest())
+	success = computed_signature == incoming_signature
+	if not success:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Signatures do not match. Please check the Incoming Key in LeafLink Settings."
+		})
+		integration_request.save(ignore_permissions=True)
+		return False
+	return True
+
+
+@frappe.whitelist(allow_guest=True)
+def sales_order_from_leaflink(**args):
+
+	#create integration request
+	integration_request = frappe.new_doc("Integration Request")
+	integration_request.update({
+		"integration_type": "Host",
+		"integration_request_service": "LeafLink",
+		"status": "Queued",
+		"data":  json.dumps(args, default=json_handler),
+		"reference_doctype": "Sales Order"
+	})
+	integration_request.save(ignore_permissions=True)
+
+	#curate data from settings for post request
+	leaflink_settings = frappe.get_cached_doc("LeafLink Settings")
+	header = {
+		leaflink_settings.get("api_key"): leaflink_settings.get("api_value")
+	}
+
+	#validate incoming keys
+	if not authenticate_leaflink_request(leaflink_settings, integration_request):
+		return
+
+	sales_order = args.get("data")
+	#exit if the action is not to create a new sales order
+	if args.get("action") != "create":
+		return
+
+	#extract customer and license data
+	customer = sales_order.get("customer")
+	customer_name = customer.get('name')
+	license_no = customer.get('license_number')
+
+	#retrieve customer_name if license_no is in record
+	if not frappe.db.exists("Customer", customer_name):
+
+		license_no = customer.get('license_number')
+		if frappe.db.exists("Compliance Info", license_no):
+
+			#find customer using license info in case name does not match
+			license_parent = frappe.get_all("Compliance License Detail",
+				filters={"license": license_no},
+				fields=['parent'])
+			if license_parent:
+				customer_name = license_parent[0].get("parent")
+		else:
+			#update integration request with message telling user that the customer needs to be created before proceeding further
+			integration_request.update({
+				"status": "Failed",
+				"error": "Customer does not exist in Bloomstack. Please create the customer and try again."
+			})
+			integration_request.save(ignore_permissions=True)
+			return
+
+	#create a sales order document with an autoset expected delivery date
+	expected_delivery_days = cint(frappe.get_value("Selling Settings", None, "expected_delivery_days")) or 7
```

Comment:
Hope the `7` is mentioned in the description as default if not specified in `expected_delivery_days` in Selling Settings